File: src/lexer.go
package src

import (
	"bytes"
	"fmt"
)

type TokenType int

const (
	TokEOF TokenType = iota
	TokIdent
	TokString
	TokLParen
	TokRParen
	TokLBrace
	TokRBrace
	TokComma
	TokColon
	TokRawScopeOpen // {!
	TokNumber       // bare integer literal, e.g. heading level
)

type Token struct {
	Type       TokenType
	Value      string
	Line, Col  int
}

type Lexer struct {
	src        []byte
	pos        int
	line, col  int
}

func NewLexer(src []byte) *Lexer {
	return &Lexer{src: src, pos: 0, line: 1, col: 1}
}

func (l *Lexer) errorf(format string, args ...interface{}) error {
	return &LexError{Line: l.line, Col: l.col, Message: fmt.Sprintf(format, args...)}
}

func (l *Lexer) peek() byte {
	if l.pos >= len(l.src) {
		return 0
	}
	return l.src[l.pos]
}

func (l *Lexer) peekAt(offset int) byte {
	if l.pos+offset >= len(l.src) {
		return 0
	}
	return l.src[l.pos+offset]
}

func (l *Lexer) advance() byte {
	c := l.src[l.pos]
	l.pos++
	if c == '\n' {
		l.line++
		l.col = 1
	} else {
		l.col++
	}
	return c
}

func (l *Lexer) skipWhitespaceAndComments() error {
	for l.pos < len(l.src) {
		c := l.peek()
		switch {
		case c == ' ' || c == '\t' || c == '\r' || c == '\n':
			l.advance()
		case c == '/' && l.peekAt(1) == '/':
			for l.pos < len(l.src) && l.peek() != '\n' {
				l.advance()
			}
		case c == '/' && l.peekAt(1) == '*':
			startLine, startCol := l.line, l.col
			l.advance()
			l.advance()
			closed := false
			for l.pos < len(l.src) {
				if l.peek() == '*' && l.peekAt(1) == '/' {
					l.advance()
					l.advance()
					closed = true
					break
				}
				l.advance()
			}
			if !closed {
				return &LexError{Line: startLine, Col: startCol, Message: "unterminated block comment"}
			}
		default:
			return nil
		}
	}
	return nil
}

func isIdentStart(c byte) bool {
	return (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z') || c == '_'
}

func isIdentCont(c byte) bool {
	return isIdentStart(c) || (c >= '0' && c <= '9')
}

// NextToken returns the next structural token, skipping whitespace/comments.
// Used for meta blocks, headers, attribute lists, and structural scanning.
func (l *Lexer) NextToken() (Token, error) {
	if err := l.skipWhitespaceAndComments(); err != nil {
		return Token{}, err
	}
	if l.pos >= len(l.src) {
		return Token{Type: TokEOF, Line: l.line, Col: l.col}, nil
	}

	startLine, startCol := l.line, l.col
	c := l.peek()

	switch {
	case c == '(':
		l.advance()
		return Token{Type: TokLParen, Value: "(", Line: startLine, Col: startCol}, nil
	case c == ')':
		l.advance()
		return Token{Type: TokRParen, Value: ")", Line: startLine, Col: startCol}, nil
	case c == '{':
		if l.peekAt(1) == '!' {
			l.advance()
			l.advance()
			return Token{Type: TokRawScopeOpen, Value: "{!", Line: startLine, Col: startCol}, nil
		}
		l.advance()
		return Token{Type: TokLBrace, Value: "{", Line: startLine, Col: startCol}, nil
	case c == '}':
		l.advance()
		return Token{Type: TokRBrace, Value: "}", Line: startLine, Col: startCol}, nil
	case c == ',':
		l.advance()
		return Token{Type: TokComma, Value: ",", Line: startLine, Col: startCol}, nil
	case c == ':':
		l.advance()
		return Token{Type: TokColon, Value: ":", Line: startLine, Col: startCol}, nil
	case c == '"':
		return l.readString()
	case isIdentStart(c):
		return l.readIdent()
	case c >= '0' && c <= '9':
		return l.readNumber()
	default:
		return Token{}, l.errorf("unexpected character %q", c)
	}
}

func (l *Lexer) readNumber() (Token, error) {
	startLine, startCol := l.line, l.col
	start := l.pos
	for l.pos < len(l.src) && l.peek() >= '0' && l.peek() <= '9' {
		l.advance()
	}
	return Token{Type: TokNumber, Value: string(l.src[start:l.pos]), Line: startLine, Col: startCol}, nil
}

func (l *Lexer) readString() (Token, error) {
	startLine, startCol := l.line, l.col
	l.advance() // opening quote
	var buf bytes.Buffer
	for {
		if l.pos >= len(l.src) {
			return Token{}, &LexError{Line: startLine, Col: startCol, Message: "unterminated string literal"}
		}
		c := l.peek()
		if c == '"' {
			l.advance()
			break
		}
		if c == '\\' && l.peekAt(1) == '"' {
			l.advance()
			buf.WriteByte(l.advance())
			continue
		}
		buf.WriteByte(l.advance())
	}
	return Token{Type: TokString, Value: buf.String(), Line: startLine, Col: startCol}, nil
}

func (l *Lexer) readIdent() (Token, error) {
	startLine, startCol := l.line, l.col
	start := l.pos
	for l.pos < len(l.src) && isIdentCont(l.peek()) {
		l.advance()
	}
	return Token{Type: TokIdent, Value: string(l.src[start:l.pos]), Line: startLine, Col: startCol}, nil
}

// ReadRawUntilBangBrace scans raw bytes verbatim starting right after "{!"
// until it finds the closing "!}" delimiter, using bytes.Index for the search.
func (l *Lexer) ReadRawUntilBangBrace() (string, int, int, error) {
	startLine, startCol := l.line, l.col
	closeIdx := bytes.Index(l.src[l.pos:], []byte("!}"))
	if closeIdx == -1 {
		return "", startLine, startCol, &LexError{Line: startLine, Col: startCol, Message: "unterminated raw scope, expected closing !}"}
	}
	raw := string(l.src[l.pos : l.pos+closeIdx])
	for i := 0; i < closeIdx+2; i++ {
		l.advance()
	}
	trimmed := trimRawBlock(raw)
	return trimmed, startLine, startCol, nil
}

// trimRawBlock removes exactly one leading newline (and its preceding
// whitespace) and trailing whitespace-only line, preserving internal
// indentation of the code.
func trimRawBlock(raw string) string {
	s := raw
	if len(s) > 0 && s[0] == '\r' {
		s = s[1:]
	}
	if len(s) > 0 && s[0] == '\n' {
		s = s[1:]
	}
	end := len(s)
	for end > 0 && (s[end-1] == ' ' || s[end-1] == '\t' || s[end-1] == '\r' || s[end-1] == '\n') {
		end--
	}
	return s[:end]
}

// PeekChar returns the raw byte at the current lexer position, skipping
// nothing. Used by prose-mode parsing to decide what construct follows a
// text run without going through the structural tokenizer.
func (l *Lexer) PeekChar() byte {
	return l.peek()
}

// braceInlineKeywords lists inline elements of the form "kw{...}" (as
// opposed to "link(...) {...}", which takes attributes first). Keeping
// this as a single shared list means AtInlineStart and ReadTextRun can
// never drift out of sync on which identifiers start an inline element.
var braceInlineKeywords = map[string]bool{
	"b":      true,
	"i":      true,
	"code":   true,
	"strike": true,
}

// AtInlineStart reports whether the lexer is currently positioned at the
// start of a brace-style inline element ("b{", "i{", "code{", "strike{")
// or "link(...) {", returning the matched keyword if so. The lexer
// position is not moved.
func (l *Lexer) AtInlineStart() (string, bool) {
	if !isIdentStart(l.peek()) {
		return "", false
	}
	ident := l.peekIdent()
	if braceInlineKeywords[ident] && l.charAfterIdent(ident) == '{' {
		return ident, true
	}
	if ident == "link" && l.charAfterIdent(ident) == '(' {
		return ident, true
	}
	return "", false
}

// ConsumeIdentOnly advances past the given identifier text (already
// confirmed present via AtInlineStart), leaving whatever follows (a '{'
// or a '(') at the current position for the caller to handle.
func (l *Lexer) ConsumeIdentOnly(ident string) {
	for i := 0; i < len(ident); i++ {
		l.advance()
	}
}

// ConsumeLBrace advances past a '{' at the current raw position, or
// returns an error if the current byte isn't '{'.
func (l *Lexer) ConsumeLBrace() error {
	if l.pos >= len(l.src) || l.peek() != '{' {
		return l.errorf("expected '{'")
	}
	l.advance()
	return nil
}

// ConsumeRBrace advances past a '}' at the current raw position, or
// returns an error if the current byte isn't '}'.
func (l *Lexer) ConsumeRBrace() error {
	if l.pos >= len(l.src) || l.peek() != '}' {
		return l.errorf("expected '}'")
	}
	l.advance()
	return nil
}

// AtRawIdent reports whether the identifier at the current raw position
// exactly matches want, without consuming anything.
func (l *Lexer) AtRawIdent(want string) bool {
	if !isIdentStart(l.peek()) {
		return false
	}
	return l.peekIdent() == want
}

// AtEOF reports whether the lexer has consumed the entire input.
func (l *Lexer) AtEOF() bool {
	return l.pos >= len(l.src)
}

// ReadRawString reads a "..."-quoted string at the current raw position
// (same escaping rules as the tokenizer's string literal) and returns its
// unescaped contents.
func (l *Lexer) ReadRawString() (string, error) {
	if l.pos >= len(l.src) || l.peek() != '"' {
		return "", l.errorf("expected string literal")
	}
	tok, err := l.readString()
	if err != nil {
		return "", err
	}
	return tok.Value, nil
}

// SkipRawSpaces advances past any spaces/tabs at the current raw position
// (not newlines — used for the tight spacing inside inline attribute lists).
func (l *Lexer) SkipRawSpaces() {
	for l.pos < len(l.src) && (l.peek() == ' ' || l.peek() == '\t') {
		l.advance()
	}
}

// ReadRawIdent reads an identifier at the current raw position.
func (l *Lexer) ReadRawIdent() string {
	return l.peekIdentAdvance()
}

func (l *Lexer) peekIdentAdvance() string {
	start := l.pos
	for l.pos < len(l.src) && isIdentCont(l.peek()) {
		l.advance()
	}
	return string(l.src[start:l.pos])
}

// ConsumeRawByte advances past a single expected byte at the current raw
// position, or returns an error naming what was expected.
func (l *Lexer) ConsumeRawByte(expected byte, what string) error {
	if l.pos >= len(l.src) || l.peek() != expected {
		return l.errorf("expected %s", what)
	}
	l.advance()
	return nil
}

// ReadBalancedBraces reads raw content up to the '}' that balances the
// '{' already consumed by the caller (depth starts at 1), tracking
// nesting depth so inner "{"/"}" pairs — e.g. a struct definition inside
// an inline code{...} span — don't prematurely end the read. The closing
// '}' itself is NOT consumed; the caller does that afterward. Strings are
// not given special treatment here: this is for raw code snippets, where
// brace characters inside string literals are rare and, if present,
// still balance correctly in the overwhelmingly common case of
// brace-structured code (C-like blocks, JSON, etc).
func (l *Lexer) ReadBalancedBraces() string {
	var buf bytes.Buffer
	depth := 1
	for l.pos < len(l.src) {
		c := l.peek()
		if c == '{' {
			depth++
		} else if c == '}' {
			depth--
			if depth == 0 {
				break
			}
		}
		buf.WriteByte(l.advance())
	}
	return buf.String()
}

// ReadTextRun reads plain text content inside a p/h block up to (but not
// including) the next inline-start marker (b{, i{, code{, strike{,
// link()) or a closing }.
func (l *Lexer) ReadTextRun() string {
	return l.readTextRunStoppingAt(nil)
}

// ReadTextRunUntilBlock behaves like ReadTextRun, but additionally stops
// right before a bare occurrence of any keyword in stopWords (e.g. "list"
// or "quote") followed by whitespace-then-'(' or whitespace-then-'{' —
// i.e. the start of a nested block construct, not just the word appearing
// in prose. This lets item/quote bodies detect a nested list{}/quote{}
// without swallowing it as plain text first.
func (l *Lexer) ReadTextRunUntilBlock(stopWords ...string) string {
	return l.readTextRunStoppingAt(stopWords)
}

func (l *Lexer) readTextRunStoppingAt(stopWords []string) string {
	var buf bytes.Buffer
	for l.pos < len(l.src) {
		c := l.peek()
		if c == '}' {
			break
		}
		if isIdentStart(c) {
			ident := l.peekIdent()
			if braceInlineKeywords[ident] && l.charAfterIdent(ident) == '{' {
				break
			}
			if ident == "link" && l.charAfterIdent(ident) == '(' {
				break
			}
			if stopWordMatches(stopWords, ident) && l.startsBlockAfterIdent(ident) {
				break
			}
		}
		buf.WriteByte(l.advance())
	}
	return buf.String()
}

func stopWordMatches(stopWords []string, ident string) bool {
	for _, w := range stopWords {
		if w == ident {
			return true
		}
	}
	return false
}

// startsBlockAfterIdent reports whether ident, at the current lexer
// position, is immediately followed (allowing spaces/tabs) by '(' or '{'
// — the shape of a block construct like "list {" or "list(ordered: true) {"
// — as opposed to the word simply appearing in running prose.
func (l *Lexer) startsBlockAfterIdent(ident string) bool {
	i := l.pos + len(ident)
	for i < len(l.src) && (l.src[i] == ' ' || l.src[i] == '\t') {
		i++
	}
	if i >= len(l.src) {
		return false
	}
	return l.src[i] == '(' || l.src[i] == '{'
}

func (l *Lexer) peekIdent() string {
	i := l.pos
	for i < len(l.src) && isIdentCont(l.src[i]) {
		i++
	}
	return string(l.src[l.pos:i])
}

func (l *Lexer) charAfterIdent(ident string) byte {
	i := l.pos + len(ident)
	if i < len(l.src) {
		return l.src[i]
	}
	return 0
}

File: src/lexer.rs
use crate::errors::LexError;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum TokenType {
    Eof,
    Ident,
    String,
    LParen,
    RParen,
    LBrace,
    RBrace,
    Comma,
    Colon,
    RawScopeOpen, // {!
    Number,       // bare integer literal
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Token {
    pub token_type: TokenType,
    pub value: String,
    pub line: usize,
    pub col: usize,
}

impl Token {
    pub fn new(token_type: TokenType, value: impl Into<String>, line: usize, col: usize) -> Self {
        Self {
            token_type,
            value: value.into(),
            line,
            col,
        }
    }
}

pub struct Lexer<'a> {
    src: &'a [u8],
    pos: usize,
    pub line: usize,
    pub col: usize,
}

impl<'a> Lexer<'a> {
    pub fn new(src: &'a [u8]) -> Self {
        Self {
            src,
            pos: 0,
            line: 1,
            col: 1,
        }
    }

    fn errorf(&self, msg: impl Into<String>) -> LexError {
        LexError::new(self.line, self.col, msg)
    }

    #[inline]
    fn peek(&self) -> u8 {
        if self.pos >= self.src.len() {
            0
        } else {
            self.src[self.pos]
        }
    }

    #[inline]
    fn peek_at(&self, offset: usize) -> u8 {
        if self.pos + offset >= self.src.len() {
            0
        } else {
            self.src[self.pos + offset]
        }
    }

    pub fn advance(&mut self) -> u8 {
        let c = self.src[self.pos];
        self.pos += 1;
        if c == b'\n' {
            self.line += 1;
            self.col = 1;
        } else {
            self.col += 1;
        }
        c
    }

    fn skip_whitespace_and_comments(&mut self) -> Result<(), LexError> {
        while self.pos < self.src.len() {
            let c = self.peek();
            match c {
                b' ' | b'\t' | b'\r' | b'\n' => {
                    self.advance();
                }
                b'/' if self.peek_at(1) == b'/' => {
                    while self.pos < self.src.len() && self.peek() != b'\n' {
                        self.advance();
                    }
                }
                b'/' if self.peek_at(1) == b'*' => {
                    let start_line = self.line;
                    let start_col = self.col;
                    self.advance();
                    self.advance();
                    let mut closed = false;
                    while self.pos < self.src.len() {
                        if self.peek() == b'*' && self.peek_at(1) == b'/' {
                            self.advance();
                            self.advance();
                            closed = true;
                            break;
                        }
                        self.advance();
                    }
                    if !closed {
                        return Err(LexError::new(
                            start_line,
                            start_col,
                            "unterminated block comment",
                        ));
                    }
                }
                _ => return Ok(()),
            }
        }
        Ok(())
    }

    pub fn next_token(&mut self) -> Result<Token, LexError> {
        self.skip_whitespace_and_comments()?;
        if self.pos >= self.src.len() {
            return Ok(Token::new(TokenType::Eof, "", self.line, self.col));
        }

        let start_line = self.line;
        let start_col = self.col;
        let c = self.peek();

        match c {
            b'(' => {
                self.advance();
                Ok(Token::new(TokenType::LParen, "(", start_line, start_col))
            }
            b')' => {
                self.advance();
                Ok(Token::new(TokenType::RParen, ")", start_line, start_col))
            }
            b'{' => {
                if self.peek_at(1) == b'!' {
                    self.advance();
                    self.advance();
                    Ok(Token::new(
                        TokenType::RawScopeOpen,
                        "{!",
                        start_line,
                        start_col,
                    ))
                } else {
                    self.advance();
                    Ok(Token::new(TokenType::LBrace, "{", start_line, start_col))
                }
            }
            b'}' => {
                self.advance();
                Ok(Token::new(TokenType::RBrace, "}", start_line, start_col))
            }
            b',' => {
                self.advance();
                Ok(Token::new(TokenType::Comma, ",", start_line, start_col))
            }
            b':' => {
                self.advance();
                Ok(Token::new(TokenType::Colon, ":", start_line, start_col))
            }
            b'"' => self.read_string(),
            _ if is_ident_start(c) => self.read_ident(),
            b'0'..=b'9' => self.read_number(),
            _ => Err(self.errorf(format!("unexpected character {:?}", c as char))),
        }
    }

    fn read_number(&mut self) -> Result<Token, LexError> {
        let start_line = self.line;
        let start_col = self.col;
        let start = self.pos;
        while self.pos < self.src.len() && self.peek().is_ascii_digit() {
            self.advance();
        }
        let val = String::from_utf8_lossy(&self.src[start..self.pos]).into_owned();
        Ok(Token::new(TokenType::Number, val, start_line, start_col))
    }

    fn read_string(&mut self) -> Result<Token, LexError> {
        let start_line = self.line;
        let start_col = self.col;
        self.advance();

        let mut buf = Vec::new();
        loop {
            if self.pos >= self.src.len() {
                return Err(LexError::new(
                    start_line,
                    start_col,
                    "unterminated string literal",
                ));
            }
            let c = self.peek();
            if c == b'"' {
                self.advance();
                break;
            }
            if c == b'\\' && self.peek_at(1) == b'"' {
                self.advance();
                buf.push(self.advance());
                continue;
            }
            buf.push(self.advance());
        }

        let val = String::from_utf8_lossy(&buf).into_owned();
        Ok(Token::new(TokenType::String, val, start_line, start_col))
    }

    fn read_ident(&mut self) -> Result<Token, LexError> {
        let start_line = self.line;
        let start_col = self.col;
        let start = self.pos;
        while self.pos < self.src.len() && is_ident_cont(self.peek()) {
            self.advance();
        }
        let val = String::from_utf8_lossy(&self.src[start..self.pos]).into_owned();
        Ok(Token::new(TokenType::Ident, val, start_line, start_col))
    }

    pub fn read_raw_until_bang_brace(&mut self) -> Result<(String, usize, usize), LexError> {
        let start_line = self.line;
        let start_col = self.col;
        let slice = &self.src[self.pos..];

        let close_idx = slice.windows(2).position(|w| w == b"!}");
        let Some(idx) = close_idx else {
            return Err(LexError::new(
                start_line,
                start_col,
                "unterminated raw scope, expected closing !}",
            ));
        };

        let raw = String::from_utf8_lossy(&self.src[self.pos..self.pos + idx]).into_owned();
        for _ in 0..(idx + 2) {
            self.advance();
        }

        let trimmed = trim_raw_block(&raw);
        Ok((trimmed, start_line, start_col))
    }

    #[inline]
    pub fn peek_char(&self) -> u8 {
        self.peek()
    }

    pub fn at_inline_start(&self) -> Option<String> {
        if !is_ident_start(self.peek()) {
            return None;
        }
        let ident = self.peek_ident();
        if is_brace_inline_keyword(&ident) && self.char_after_ident(&ident) == b'{' {
            return Some(ident);
        }
        if (ident == "link" || ident == "fn") && self.char_after_ident(&ident) == b'(' {
            return Some(ident);
        }
        None
    }

    pub fn consume_ident_only(&mut self, ident: &str) {
        for _ in 0..ident.len() {
            self.advance();
        }
    }

    pub fn consume_lbrace(&mut self) -> Result<(), LexError> {
        if self.pos >= self.src.len() || self.peek() != b'{' {
            return Err(self.errorf("expected '{'"));
        }
        self.advance();
        Ok(())
    }

    pub fn consume_rbrace(&mut self) -> Result<(), LexError> {
        if self.pos >= self.src.len() || self.peek() != b'}' {
            return Err(self.errorf("expected '}'"));
        }
        self.advance();
        Ok(())
    }

    pub fn at_raw_ident(&self, want: &str) -> bool {
        if !is_ident_start(self.peek()) {
            return false;
        }
        self.peek_ident() == want
    }

    pub fn at_eof(&self) -> bool {
        self.pos >= self.src.len()
    }

    pub fn read_raw_string(&mut self) -> Result<String, LexError> {
        if self.pos >= self.src.len() || self.peek() != b'"' {
            return Err(self.errorf("expected string literal"));
        }
        let tok = self.read_string()?;
        Ok(tok.value)
    }

    pub fn skip_raw_spaces(&mut self) {
        while self.pos < self.src.len() && (self.peek() == b' ' || self.peek() == b'\t') {
            self.advance();
        }
    }

    /// Reads raw identifier chars, including hyphens (-) for footnotes and attributes.
    pub fn read_raw_ident(&mut self) -> String {
        let start = self.pos;
        while self.pos < self.src.len() && (is_ident_cont(self.peek()) || self.peek() == b'-') {
            self.advance();
        }
        String::from_utf8_lossy(&self.src[start..self.pos]).into_owned()
    }

    pub fn consume_raw_byte(&mut self, expected: u8, what: &str) -> Result<(), LexError> {
        if self.pos >= self.src.len() || self.peek() != expected {
            return Err(self.errorf(format!("expected {what}")));
        }
        self.advance();
        Ok(())
    }

    pub fn read_balanced_braces(&mut self) -> String {
        let mut buf = Vec::new();
        let mut depth = 1;
        while self.pos < self.src.len() {
            let c = self.peek();
            if c == b'{' {
                depth += 1;
            } else if c == b'}' {
                depth -= 1;
                if depth == 0 {
                    break;
                }
            }
            buf.push(self.advance());
        }
        String::from_utf8_lossy(&buf).into_owned()
    }

    pub fn read_text_run(&mut self) -> String {
        self.read_text_run_stopping_at(&[])
    }

    pub fn read_text_run_until_block(&mut self, stop_words: &[&str]) -> String {
        self.read_text_run_stopping_at(stop_words)
    }

    fn read_text_run_stopping_at(&mut self, stop_words: &[&str]) -> String {
        let mut buf = Vec::new();
        while self.pos < self.src.len() {
            let c = self.peek();
            if c == b'}' {
                break;
            }
            if is_ident_start(c) {
                let ident = self.peek_ident();
                if is_brace_inline_keyword(&ident) && self.char_after_ident(&ident) == b'{' {
                    break;
                }
                if (ident == "link" || ident == "fn") && self.char_after_ident(&ident) == b'(' {
                    break;
                }
                if stop_words.contains(&ident.as_str()) && self.starts_block_after_ident(&ident) {
                    break;
                }
            }
            buf.push(self.advance());
        }
        String::from_utf8_lossy(&buf).into_owned()
    }

    fn starts_block_after_ident(&self, ident: &str) -> bool {
        let mut i = self.pos + ident.len();
        while i < self.src.len() && (self.src[i] == b' ' || self.src[i] == b'\t') {
            i += 1;
        }
        if i >= self.src.len() {
            return false;
        }
        self.src[i] == b'(' || self.src[i] == b'{'
    }

    fn peek_ident(&self) -> String {
        let mut i = self.pos;
        while i < self.src.len() && is_ident_cont(self.src[i]) {
            i += 1;
        }
        String::from_utf8_lossy(&self.src[self.pos..i]).into_owned()
    }

    fn char_after_ident(&self, ident: &str) -> u8 {
        let i = self.pos + ident.len();
        if i < self.src.len() {
            self.src[i]
        } else {
            0
        }
    }
}

#[inline]
fn is_ident_start(c: u8) -> bool {
    c.is_ascii_alphabetic() || c == b'_'
}

#[inline]
fn is_ident_cont(c: u8) -> bool {
    is_ident_start(c) || c.is_ascii_digit()
}

#[inline]
fn is_brace_inline_keyword(s: &str) -> bool {
    matches!(s, "b" | "i" | "code" | "strike" | "m")
}

fn trim_raw_block(raw: &str) -> String {
    let mut s = raw;
    if let Some(idx) = s.find('\n') {
        let prefix = &s[..idx];
        if prefix
            .bytes()
            .all(|b| b == b' ' || b == b'\t' || b == b'\r')
        {
            s = &s[idx + 1..];
        }
    }
    s.trim_end_matches([' ', '\t', '\r', '\n']).to_string()
}
